decryptor/postgresql/data_encoder.go

```diff
@@ -0,0 +1,249 @@
+package postgresql
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"github.com/cossacklabs/acra/decryptor/base"
+	"github.com/cossacklabs/acra/encryptor"
+	"github.com/cossacklabs/acra/encryptor/config"
+	common2 "github.com/cossacklabs/acra/encryptor/config/common"
+	"github.com/cossacklabs/acra/logging"
+	"github.com/cossacklabs/acra/utils"
+	"github.com/sirupsen/logrus"
+	"strconv"
+)
+
+// PgSQLDataEncoderProcessor implements processor and encode binary/text values before sending to app
+type PgSQLDataEncoderProcessor struct{}
+
+// NewPgSQLDataEncoderProcessor return new data encoder to text/binary format
+func NewPgSQLDataEncoderProcessor() (*PgSQLDataEncoderProcessor, error) {
+	return &PgSQLDataEncoderProcessor{}, nil
+}
+
+// ID return name of processor
+func (p *PgSQLDataEncoderProcessor) ID() string {
+	return "PgSQLDataEncoderProcessor"
+}
+
+// here we process encryption/tokenization results before send it to a client
+// acra decrypts or de-tokenize SQL literals, so we should convert string SQL literals to binary format
+// if client expects int, then parse INT literals and convert to binary 4/8 byte format
+// if expects bytes, then pass as is
+// if expects string, then leave as is if it is valid string or encode to hex
+// if it is encrypted data then we return default values or as is if applicable (binary data)
+func (p *PgSQLDataEncoderProcessor) encodeBinary(ctx context.Context, data []byte, setting config.ColumnEncryptionSetting, columnInfo base.ColumnInfo, logger *logrus.Entry) (context.Context, []byte, error) {
+	if len(data) == 0 {
+		return ctx, data, nil
+	}
+	switch setting.GetEncryptedDataType() {
+	case common2.EncryptedType_String:
+		if !base.IsDecryptedFromContext(ctx) {
+			if newVal := setting.GetDefaultDataValue(); newVal != nil {
+				return ctx, []byte(*newVal), nil
+			}
+		}
+		return ctx, data, nil
+	case common2.EncryptedType_Bytes:
+		if !base.IsDecryptedFromContext(ctx) {
+			if newVal := setting.GetDefaultDataValue(); newVal != nil {
+				binValue, err := base64.StdEncoding.DecodeString(*newVal)
+				if err == nil {
+					return ctx, binValue, nil
+				}
+				logger.WithError(err).Errorln("Can't decode base64 default value")
+			}
+		}
+		return ctx, data, nil
+	case common2.EncryptedType_Int32, common2.EncryptedType_Int64:
+		size := 8
+		if setting.GetEncryptedDataType() == common2.EncryptedType_Int32 {
+			size = 4
+		}
+		// convert back from text to binary
+		value, err := strconv.ParseInt(string(data), 10, 64)
```

Comment:
Small question: in all other cases we first check that value is indeed decrypted (by calling `base.IsDecryptedFromContext(ctx)`), and if not, we fallback to returning default values. But in this case we check decryption by parsing the number. Does it have some special meaning or reason?

Reply:
the main reason is that any encrypted blob has size > len('9223372036854775807') = 19. Our acrablock has 90+ bytes of overhead. Acrastructs even more. So no one supported ciphertext cannot be valid integer string literal. So there is enough to check that it's valid valid integer string literal.